CreateCategoryServiceTest Red, Green, Blue 완료

xpmxf4 opened the pull request:
드릅게 할게 많더라구요,,,

xpmxf4 committed:
🚦 test: CategoryRepository Test 작성

xpmxf4 committed:
⚙️  chore: test 안 category 폴더 생성
추후 선우님의 폴더구조 변경 사항에 따라 한번 더 달라질 예정

xpmxf4 committed:
🔴 red-test: CreateCategoryServiceTest Red

xpmxf4 committed:
🟢 green-test: CreateCategoryServiceTest Green

xpmxf4 committed:
🔵 blue-test: createCategory 내 부모 id 체크 추상화

ch4570 reviewed (commented):
parentId는 AutoIncrement로 지정되는 값이 아닌지 궁금합니다.
왜 -1 인지 체크하는지 코드 작성 의도가 궁금해요:)

ch4570 commented on src/test/java/com/member/homework/service/category/CreateCategoryServiceTest.java:
```diff
@@ -38,25 +38,16 @@ void adminCanCreateNewCategory() {
+                = new CreateCategoryServiceRequest("name", "desc", 10L);
 
         // when - 동작, then
         assertThatThrownBy(() -> createCategoryService.createCategory(createCategoryServiceRequest))
```
assertThatThrownBy() 뒤에 예외 메시지와 인스턴스 체크하는 코드는 없는 걸까요??

ch4570 commented on src/test/java/com/member/homework/service/category/CreateCategoryServiceTest.java:
```diff
@@ -0,0 +1,67 @@
+
+        // when - 동작
+        createCategoryService.createCategory(createCategoryServiceRequest);
+        categoryRepository.flush();
```
왜 Category Repository에서 push를 해주셨는지 알고 싶어요:)

트랜잭션이 끝날 때, DB에 쿼리가 나가기 때문에 쿼리가 실제 나가도록 하려고 하신거 같기도 한데
결국 한 번 저장하면 영속성 컨텍스트에 코드가 남아있기 때문에 1차 캐시의 내용이 나올텐데 어떤 부분을 테스트 하려고
하신건지 알고 싶어요!

ch4570 commented on src/test/java/com/member/homework/repository/CategoryRepository.java:
```diff
@@ -0,0 +1,11 @@
+@Repository
+public interface CategoryRepository extends JpaRepository<Category, Long> {
+
+    Category findByName(String name);
```
Null-Safe 하도록 Optional을 사용하는 것이 더 안전하지 않을까요?

ch4570 commented on src/test/java/com/member/homework/repository/CategoryRepositoryTest.java:
```diff
@@ -0,0 +1,106 @@
+    // 이게 필요한 테스트일까?
+    @Test
+    @DisplayName("새로운 카테고리를 생성할 수 있다.")
+    void adminCreatesNewCategorySuccessfully() {
```
저장 테스트도 저는 필요하다고 생각해요:)

왜냐하면 저장할 때 JPA 엔티티의 문법이나 다른 문제 때문에 저장이 제대로 되지 않을 수도 있을 거 같아서요:)
카테고리 엔티티가 현재는 단독 엔티티지만, 다른 엔티티와 연관 관계를 가지게 된다면 그에 따른 테스트가 추가 될 수도
있을 거 같아요

ch4570 commented on src/test/java/com/member/homework/repository/CategoryRepositoryTest.java:
```diff
@@ -0,0 +1,106 @@
+    @DisplayName("같은 이름의 카테고리는 생성할 수 없다.")
+    void cannotMakeCategoryWithExistingName() {
+        // given - 상황 만들기
+        Category existingCategory = Category.of("해외축구갤러리", "해축갤 만세", null);
```
existingCategory와 newCategory 보다는 아래와 같은 네이밍이 더 직관적이지 않을까 제안드려봐요:)

1. 이미 있는 카테고리 -> newCategory(새로 저장하는 카테고리니까)
2. 중복 저장 시도하는 카테고리 -> duplicateCategory(새로 저장한 카테고리에 뒤이어서 중복 저장을 시도하니까)

ch4570 commented on src/test/java/com/member/homework/repository/CategoryRepositoryTest.java:
```diff
@@ -0,0 +1,106 @@
+    @DisplayName("특정 카테고리를 삭제할 수 있다.")
+    void adminCanDeleteSpecificCategory() {
+        // given - 상황 만들기
+        Category categoryToBeDeleted = Category.of("1", "1", null);
```
지워질 예정인 카테고리라도 조금 더 단순한 이름으로 표현해보면 어떨까요?
만약 제가 이 코드를 작성한다면 category 라는 단순한 이름을 쓸 것 같아요

ch4570 commented on src/test/java/com/member/homework/repository/CategoryRepositoryTest.java:
```diff
@@ -0,0 +1,106 @@
+
+        // when - 동작
+        categoryRepository.delete(categoryToBeDeleted);
+        Category unexistingCategory = categoryRepository.findByName("1");
```
위에서 category 라는 이름을 쓰고, 이 쿼리 메서드를 existsByName()으로 이름을 변경한다면
Boolean 타입의 결과를 얻으니까 isFalse()로 검증하면 될 것 같아요

ch4570 commented on src/test/java/com/member/homework/repository/CategoryRepositoryTest.java:
```diff
@@ -0,0 +1,106 @@
+        Category unexistingCategory = categoryRepository.findByName("1");
+
+        // then - 검증
+        assertThat(unexistingCategory).isNull();
```
또 Null을 직접 다루는 것은 NPE의 위험이 크기 때문에 Optional로 풀어냈으면 좋겠어요

Commit:
🟢 Test-Green: AdminController Test 작성 완료

Commit:
🟢 Test-Green: 회원 전체 조회 테스트 까지 완료

xpmxf4 committed:
🔴 red-test: LoadCategoryServiceTest Red
1. 카테고리 조회 성공시, 리스트를 가져온다.
2. 카테고리가 없으면 빈 리스트를 반환한다.

xpmxf4 committed:
🟢 green-test: LoadCategoryServiceTest Green

xpmxf4 committed:
💥 Conflict : git conflict 해결

xpmxf4 committed:
🔴 red-test: RemoveCategoryServiceTest Red
1. 아무 게시글이 없는 카테고리의 경우, 삭제할 수 있다.
2. 존재하지 않는 카테고리의 ID 로 삭제할 수 없다.
3. 게시글이 존재하는 카테고리는 삭제가 되지 않는다.

xpmxf4 committed:
🟢 green-test: RemoveCategoryServiceTest Green

xpmxf4 committed:
🔴 red-test: ModifyCategoryService Test Red

xpmxf4 committed:
🟢 green-test: ModifyCategoryService Test Green

xpmxf4 committed:
🔵 blue-test: ModifyCategory blue
존재하지 않는 부모로 변경할 때도 안되는 법 검증

xpmxf4 committed:
⚙️  chore: 오타수정

xpmxf4 commented:
만약 부모를 안 바꿀거면 명시적으로 -1 이라는 값으로 넘어온다고 가정했습니다!

xpmxf4 committed:
🔵 blue-test: ModifyCategoryService Blue
존재하지 않는 부모 id 를 조회하는 경우
메세지를 '없는 부모 id' 조회한다고 구체적으로 명시함
